include/dlaf/eigensolver/tridiag_solver/merge.h

```diff
@@ -433,6 +437,48 @@ auto applyDeflation(const SizeType i_begin, const SizeType i_end, RhoSender&& rh
          ex::ensure_started();
 }
 
+void initSortByColTypeIndex(SizeType i_begin, SizeType i_end,
+                            Matrix<const ColType, Device::CPU>& coltypes,
+                            Matrix<SizeType, Device::CPU>& index) {
+  namespace ex = pika::execution::experimental;
+  namespace di = dlaf::internal;
+
+  const SizeType n = problemSize(i_begin, i_end, index.distribution());
+  TileCollector tc{i_begin, i_end};
+
+  ex::start_detached(
+      ex::when_all(ex::when_all_vector(tc.read(coltypes)), ex::when_all_vector(tc.readwrite(index))) |
+      di::transform(di::Policy<Backend::MC>{}, [n](auto&& coltype_tiles, auto&& index_tiles) {
+        auto coltype_index = [](const ColType coltype) -> std::size_t {
+          switch (coltype) {
+            case ColType::UpperHalf:
+              return 0;
+            case ColType::Dense:
+              return 1;
+            case ColType::LowerHalf:
+              return 2;
+            case ColType::Deflated:
+              return 3;
+          }
+          return DLAF_UNREACHABLE(std::size_t);
+        };
```

Comment:
I opted for this instead of adding values to the enum class because it represents an important implementation detail of the algorithm, and not of the enum class itself (constraint: they have to be contiguous, they have to start from 0 and they should represent the exact order how we want columns to be sorted).

If you feel it is better the other way, I can easily change this.

Reply:
For me this is ok, though I admit it took me a moment to understand what was going on here. You're essentially using the enum values as indices into an array. It'd be interesting if the compiler anyways sees through that `coltype_index` function sees that the mapping is `0` to `0`, `1` to `1` etc. "Interesting" but I'm not asking you to check.

No action required here (for me).

Reply:
It is exactly a LUT, but I didn't find a better way to create a static one (apart specifying values for `ColType` values in the `enum class` definition.

Note: the LUT creates an order that is slightly different from how they are defined in the `enum class`. 

```cpp
{Upper, Dense, Lower, Deflated} // as per LUT with switch
{Upper, Lower, Dense, Deflated} // as per enum class ColType definition
```

https://github.com/eth-cscs/DLA-Future/blob/7752e7b236951daab952c733413cf3d5175ae942/include/dlaf/eigensolver/tridiag_solver/coltype.h#L17-L22

This is part of why I didn't want to put an implementation detail in the `enum class` definition...it looked too risky to me (e.g. someone might decide to reorder/change the values in the definition).

Reply:
I feel like I'mmissing something obvious, but why is the order different between the enum and your indices? I understand that they're can go out of sync and that's ok, but is there a good reason for the particular order?

Reply:
No, there is no reason for the current order of declaration in the enum. And that's why I didn't want to change that: there is no reason to order them differently, apart from an algorithmic detail.

I agree that can be changed for following the algorithmic requirement, but at that point I would have to find a way to constrain it/check it somehow (otherwise an "involuntary" change to the order would end up in a broken algorithm).

Reply:
Thanks @albestro. I was rather asking the opposite: why is the order in the algorithm not changed to be the same as in the enum? You're hinting at some algorithmic constraint that results in the current order, and if there is such a constraint it should probably be added to a comment in the implementation.

Reply:
This PR sorts the eigenvectors in order to make the matrix storing them have a "nicer shape", so that general multiplication can be done more efficiently. Indeed, it sorts them as in

https://github.com/eth-cscs/DLA-Future/blob/3d21dc900bdb4fb5ce30ebbacded3230150f6623/include/dlaf/eigensolver/tridiag_solver/merge.h#L759-L763

so that later we will be able to perform the multiplication (deflated do not take part in this) as two sub-multiplication (top left and bottom right sub-matrices) skipping the zeros part, reducing the number of operations needed.

Reply:
Ok, I still feel like some comment at this point could be good, but the link to the explanation is already good enough for me. Thanks!